### PR TITLE
Himanshu/sync news

### PR DESCRIPTION
**Pull request by HGoyal09**
Merging this pr will - 
Set up a cron job to pull in stock news 
Use the stock news on the front end 
Sync real time stock data 
image

### CODE REVIEW

File: server/crons/stock-news.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+}, {})
+
+const newsArticleExists = (stock_news, article) => (
+    stock_news.find(news => news['publishedAt'] === article.publishedAt)
+)
+
+const getStockNewsFromApi = async ({name}) => (
+    await newsapi.v2.topHeadlines({
+        q: `${name}`,
```
Comment (raghavthakur) [inline server/crons/stock-news.js:48]: What does q stand for?

File: server/dal/stock-info-old.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+var Schema = mongoose.Schema
+
+var stockInfoSchema = new Schema({
+    company_name: String,
+    ticker: String,
+    current_price: Number,
+    previous_close: Number,
+    open: Number,
+    last_price: Number,
```
Comment (raghavthakur) [inline server/dal/stock-info-old.js]: I think we decided not to use last_price as it's the same as previous_close. So we should update the schema to have only previous_close and remove last_price.

File: server/dal/stock-info-old.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    change: Number,
+    change_percent: String,
+    currency: String,
+    volume: String,
+    pe_ratio: Number,
+    shares_owned: Number,
+    avg_vol: String,
+    market_cap: String,
```
Comment (raghavthakur) [inline server/dal/stock-info-old.js]: Have we decided on using strings for large number values? I think we talked a little about using the number type in the backend and then have the frontend truncate the number for viewing format purposes.

File: server/dal/stock-markets.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+var realtimeStockInfoSchema = new Schema({
+    market_name: String,
+    market_id: String,
+    stock_name: String,
+    ticker_id: String,
+    stock_id: String,
+    stock_details: {
+        timestamp: Date,
+        previous_close: Number,
```
Comment (raghavthakur) [inline server/dal/stock-markets.js]: You only used previous_close here which is good, but I think forgot to remove last_price from other areas.

File: server/data/stock-market.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            ticker: 'FB',
+            name: "Facebook"
+        },
+        {
+            ticker: 'TWTR',
+            name: "Twitter"
+        },
+        {
+            ticker: 'UBER',
+            name: "Uber Technologies Inc"
```
Comment (raghavthakur) [inline server/data/stock-market.js]: If you're planning on including "Inc" here rather than just "Uber" then I think you should be consistent and use it everywhere it's required. Or just use "Uber" rather than the full legal name of the companies. Just a minor consistency issue.

File: server/data/stock-market.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        }
+    ]
+}]
+
+for (const data of MARKET_DATA) {
+    const {market_name, stocks} = data
+
+    stockMarketInfo.find({market_name: market_name}).then((doc) => {
+        if (doc.length === 0) {
+            var stockMarketData = new stockMarketInfo(data)
```
Comment (raghavthakur) [inline server/data/stock-market.js]: Shouldn't this be let, because it's a block scoped declaration rather than function scoped used for var

File: server/routes/stock-news.js [hunk 7]
```diff
@@ -1,4 +1,5 @@
 var express = require('express');
+const {stockNewsInfo} = require("../dal/stock-markets");
```
Comment (raghavthakur) [inline server/routes/stock-news.js:2]: What's does the dal folder naming mean?

**Review by raghavthakur — CHANGES_REQUESTED** [review]

**Review by raghavthakur — APPROVED** [review]

File: server/crons/quarterly-stock-data.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
+const dotenv = require('dotenv');
+dotenv.config();
+const {stockMarketInfo} = require("../dal/stock-markets");
+const NewsAPI = require('newsapi');
+require('dotenv').config({path: '../.env'});
```
Comment (weiyoung) [inline server/crons/quarterly-stock-data.js:3]: small issue but why is this line here and not say, after line 2?

File: server/crons/quarterly-stock-data.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+const stockDataApiKey = new NewsAPI(process.env.STOCK_DATA_API_KEY);
+
+const cron = require('node-cron')
+const axios = require("axios");
+const {quarterlyStockInfo} = require("../dal/stock-markets");
```
Comment (weiyoung) [inline server/crons/quarterly-stock-data.js]: can you merge this line and line 3?

File: server/crons/realtime-stock-data.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
+const dotenv = require('dotenv');
+dotenv.config();
+const {stockMarketInfo} = require("../dal/stock-markets");
+const NewsAPI = require('newsapi');
+require('dotenv').config({path: '../.env'});
+const stockDataApiKey = new NewsAPI(process.env.STOCK_DATA_API_KEY);
+
+const cron = require('node-cron')
+const axios = require("axios");
+const {realtimeStockInfo} = require("../dal/stock-markets");
```
Comment (weiyoung) [inline server/crons/realtime-stock-data.js]: this as well, merge with line 3?

File: server/crons/stock-news.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
+const dotenv = require('dotenv');
+dotenv.config();
+const {stockMarketInfo} = require("../dal/stock-markets");
+const NewsAPI = require('newsapi');
+require('dotenv').config({path: '../.env'});
+const newsapi = new NewsAPI(process.env.NEWS_API_KEY);
+
+const cron = require('node-cron')
+const {stockNewsInfo} = require("../dal/stock-markets");
```
Comment (weiyoung) [inline server/crons/stock-news.js]: this too, line 3?

File: server/crons/stock-news.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+cron.schedule("0 0 0 * * *", async () => {
+    const doc = await stockMarketInfo.find({})
+    for (let market_data of doc) {
+        for (let stock_data of market_data.stocks) {
+            const result = await stockNewsInfo.findOne({ticker_id: stock_data.ticker})
+            const news_data = await getStockNewsFromApi(stock_data)
+            const {articles} = news_data
```
Comment (weiyoung) [inline server/crons/stock-news.js]: can we remove line 17? aka replace the variable `news_data` with {articles} itself

File: server/routes/stock-details.js [hunk 13]
```diff
@@ -20,49 +21,15 @@ const defaultDetails = {
...
-        fifty_two_week_high: 72.19
-    }
-]
 
 /* GET stock details. */
 router.get('/:ticker', function (req, res) {
-    const details = stockInfo.filter((stock) => stock.ticker === req.params.ticker);
-    res.send(details.length === 1 ? details[0] : defaultDetails)
+    stockInfo.find().then(stockList => {
+        const details = stockList.filter((stock) => stock.ticker === req.params.ticker);
```
Comment (weiyoung) [inline server/routes/stock-details.js]: instead of using find() with an empty query, could we query with req.params.ticker instead, so we don't request for the whole stocklist? would save us network weight

File: server/routes/stock-prediction.js [hunk 14]
```diff
@@ -11,30 +11,14 @@ const defaultPredictions = {
...
-    }
-]
-
-
 /* GET prediction details. */
 router.get('/:ticker', function (req, res) {
-    const details = stockPredictions.filter((stock) => stock.ticker === req.params.ticker);
-    res.send(details.length === 1 ? details[0] : defaultPredictions)
+    stockPrediction.find().then(stockList => {
+        const details = stockList.filter((stock) => stock.ticker === req.params.ticker);
```
Comment (weiyoung) [inline server/routes/stock-prediction.js]: same thing here, can we find() with the specific ticker that we want?

**Review by weiyoung — APPROVED** [review]
